Core/src/RequestHandler.php

```diff
@@ -0,0 +1,154 @@
+    use RequestCallerTrait;
+
+    /**
+     * @var Serializer
+     */
+    private $serializer;
+
+    private $gapics;
+
+    /**
+     * @param array $config
+     */
+    public function __construct(
+        Serializer $serializer,
+        array $gapicClasses,
+        array $clientConfig = []
+    ) {
+        //@codeCoverageIgnoreStart
+        $this->serializer = $serializer;
+        $clientConfig['serializer'] = $serializer;
+
+        $this->setRequestWrapper(new GapicRequestWrapper($clientConfig));
+
+        // Adds some defaults
+        // gccl needs to be present for handwritten clients
+        $clientConfig += [
+            'libName' => 'gccl',
+            'transport' => $this->getDefaultTransport(),
+            'emulatorHost' => null
+        ];
+
+        if ((bool) $clientConfig['emulatorHost']) {
+            $emulatorConfig = $this->emulatorGapicConfig($clientConfig['emulatorHost']);
+            $clientConfig = array_merge(
+                $clientConfig,
+                $emulatorConfig
+            );
+        }
+        //@codeCoverageIgnoreEnd
+        
+        // Initialize the Gapic classes and store them in memory
+        $this->gapics = [];
+        foreach ($gapicClasses as $cls) {
+            $this->gapics[$cls] = new $cls($clientConfig);
+        }
+    }
+
+    /**
+     * Helper function that forwards the request to a gapic client obj.
+     *
+     * @param $gapicClass The request will be forwarded to this GAPIC class.
+     * @param $method This method needs to be called on the gapic obj.
+     * @param $requiredArgs The positional arguments to be passed on the $method
+     * @param $args The optional args.
+     *
+     * If a GAPIC class is provided as the first argument,
+     * then the GAPIC object already stored in memory is used.
+     * If a GAPIC object is supplied,then we use the object as is.
+     * This is useful to override the GAPIC object used for one specific request.
+     */
+    public function sendRequest(
+        $gapicClass,
+        string $method,
+        array $requiredArgs,
+        array $optionalArgs,
+        bool $whitelisted = false
+    ) {
+
+        $allArgs = $requiredArgs;
+
+        // we send the optional args in the end, because everything before that is
+        // passed on the the `$method` as a positional argument
+        $allArgs[] = $optionalArgs;
+
+        $gapicObj = $this->getGapicObject($gapicClass);
+
+        // TODO: check how can we simplify the use of $whitelisted
+        return $this->send([$gapicObj, $method], $allArgs, $whitelisted);
+    }
+
+    /**
+     * Returns the current serializer instance.
+     *
+     * @return Serializer
+     */
+    public function getSerializer()
+    {
+        return $this->serializer;
+    }
+
+    /**
+     * Setter for the serializer.
+     *
+     * @return void
+     */
+    public function setSerializer(Serializer $serializer)
+    {
+        $this->serializer = $serializer;
+    }
```

Comment:
Why do we allow the setting of the serializer after the object is created? This will only cause a divergence between the serializer in this class, and the one in `GapicRequestWrapper`, which will open the door for bugs and unexpected behavior. As far as I know, there's no reason to allow changing the serializer after we've created the `RequestHander`.

Reply:
Removed